File: skills/image-generation/scripts/mj_imagine_client.py
#!/usr/bin/env python3
import argparse
import base64
import json
import mimetypes
import os
import sys
import time
import urllib.error
import urllib.request
from typing import Any, Dict, List, Optional

DEFAULT_BASE_URL = "https://api.bltcy.top"
DEFAULT_MJ_VERSION = "7"
CONFIG_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "api_config.json")


def load_config(config_path: str = CONFIG_FILE) -> Dict[str, Any]:
    if not os.path.exists(config_path):
        raise RuntimeError(f"Config file not found: {config_path}")
    try:
        with open(config_path, "r", encoding="utf-8") as f:
            config = json.load(f)
    except json.JSONDecodeError as e:
        raise RuntimeError(f"Invalid JSON in config file: {config_path}") from e
    if not isinstance(config, dict):
        raise RuntimeError(f"Config file must be a JSON object: {config_path}")
    return config


def load_api_key(config: Dict[str, Any]) -> str:
    api_key = str(config.get("api_key") or "").strip()
    if not api_key:
        raise RuntimeError(f"Missing 'api_key' in config file: {CONFIG_FILE}")
    return api_key


def resolve_base_url(args: argparse.Namespace, config: Dict[str, Any]) -> str:
    value = args.base_url or config.get("mj_base_url") or config.get("base_url") or DEFAULT_BASE_URL
    value = str(value).strip().rstrip("/")
    if not value:
        raise RuntimeError("Base URL is empty")
    return value


def resolve_route_prefix(args: argparse.Namespace, config: Dict[str, Any]) -> str:
    raw = args.route_prefix
    if raw is None:
        raw = config.get("mj_route_prefix", "fast")
    raw = str(raw).strip()
    if not raw:
        return ""
    return raw.strip("/")


def build_submit_url(base_url: str, route_prefix: str) -> str:
    if route_prefix:
        return f"{base_url}/{route_prefix}/mj/submit/imagine"
    return f"{base_url}/mj/submit/imagine"


def build_fetch_url(base_url: str, route_prefix: str, task_id: str) -> str:
    if route_prefix:
        return f"{base_url}/{route_prefix}/mj/task/{task_id}/fetch"
    return f"{base_url}/mj/task/{task_id}/fetch"


def encode_image_to_data_uri(image_path: str) -> str:
    if not os.path.exists(image_path):
        raise FileNotFoundError(f"Image file not found: {image_path}")
    mime_type, _ = mimetypes.guess_type(image_path)
    if not mime_type:
        mime_type = "image/png"
    with open(image_path, "rb") as f:
        encoded = base64.b64encode(f.read()).decode("utf-8")
    return f"data:{mime_type};base64,{encoded}"


def build_base64_array(image_inputs: Optional[List[str]]) -> List[str]:
    if not image_inputs:
        return []
    items: List[str] = []
    for item in image_inputs:
        text = item.strip()
        if not text:
            continue
        if text.startswith("data:image/"):
            items.append(text)
            continue
        if text.startswith("http://") or text.startswith("https://"):
            # The API document focuses on base64Array. Keep URL passthrough
            # for compatibility with providers that accept remote images.
            items.append(text)
            continue
        items.append(encode_image_to_data_uri(text))
    return items


def normalize_prompt(prompt: str, default_version: str = DEFAULT_MJ_VERSION) -> str:
    text = prompt.strip()
    if not text:
        return text
    lower = f" {text.lower()} "
    if " --v " in lower or " --version " in lower:
        return text
    return f"{text} --v {default_version}"


def request_json(method: str, url: str, api_key: str, body_obj: Optional[Dict[str, Any]], timeout: int) -> Dict[str, Any]:
    body = None if body_obj is None else json.dumps(body_obj).encode("utf-8")
    req = urllib.request.Request(
        url=url,
        data=body,
        method=method,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw = resp.read().decode("utf-8")
            return json.loads(raw)
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"HTTP {e.code}: {err_body}") from e
    except urllib.error.URLError as e:
        raise RuntimeError(f"Network error: {e}") from e


def submit_imagine(api_key: str, submit_url: str, prompt: str, base64_array: List[str], notify_hook: str, timeout: int) -> Dict[str, Any]:
    payload: Dict[str, Any] = {
        "prompt": prompt,
        "base64Array": base64_array,
    }
    if notify_hook:
        payload["notifyHook"] = notify_hook
    return request_json("POST", submit_url, api_key, payload, timeout)


def poll_task(
    api_key: str,
    fetch_url: str,
    timeout: int,
    poll_interval: float,
    poll_timeout: int,
) -> Dict[str, Any]:
    deadline = time.time() + poll_timeout
    last: Dict[str, Any] = {}
    while time.time() < deadline:
        last = request_json("GET", fetch_url, api_key, None, timeout)
        status = str(last.get("status") or "").upper()
        progress = str(last.get("progress") or "")

        if status in {"SUCCESS", "FAILURE", "CANCEL"}:
            return last
        if progress == "100%":
            return last
        time.sleep(max(0.2, poll_interval))
    raise TimeoutError(f"Polling timed out after {poll_timeout}s")


def download_image(url: str, output_path: str, timeout: int) -> None:
    req = urllib.request.Request(url=url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        content = resp.read()
    os.makedirs(os.path.dirname(os.path.abspath(output_path)), exist_ok=True)
    with open(output_path, "wb") as f:
        f.write(content)


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Midjourney Imagine client (文生图 + 文图生图 via base64Array)"
    )
    parser.add_argument("--prompt", required=True, help="MJ prompt")
    parser.add_argument(
        "--image-path",
        nargs="*",
        default=[],
        help="Optional reference image paths/URLs/data-URIs for 文图生图",
    )
    parser.add_argument("--base-url", default="", help="Override API base URL")
    parser.add_argument(
        "--route-prefix",
        default=None,
        help="Route prefix before /mj, e.g. fast, mj-turbo, mj-relax; empty for direct /mj",
    )
    parser.add_argument("--notify-hook", default="", help="Optional callback URL (notifyHook)")
    parser.add_argument("--timeout", type=int, default=120, help="Single request timeout in seconds")
    parser.add_argument("--no-poll", action="store_true", help="Only submit task and exit")
    parser.add_argument("--poll-interval", type=float, default=3.0, help="Polling interval in seconds")
    parser.add_argument("--poll-timeout", type=int, default=600, help="Polling timeout in seconds")
    parser.add_argument("--download", default="", help="Download final image when task succeeds")

    args = parser.parse_args()

    try:
        config = load_config(CONFIG_FILE)
        api_key = load_api_key(config)
        base_url = resolve_base_url(args, config)
        route_prefix = resolve_route_prefix(args, config)
    except Exception as e:
        print(f"Config error: {e}", file=sys.stderr)
        return 2

    try:
        base64_array = build_base64_array(args.image_path)
    except Exception as e:
        print(f"Image error: {e}", file=sys.stderr)
        return 2

    submit_url = build_submit_url(base_url, route_prefix)
    final_prompt = normalize_prompt(args.prompt)

    print(f"Submit URL: {submit_url}", file=sys.stderr)
    print(f"Mode: {'文图生图' if base64_array else '文生图'}", file=sys.stderr)
    if final_prompt != args.prompt.strip():
        print(f"Auto append default MJ version: --v {DEFAULT_MJ_VERSION}", file=sys.stderr)

    try:
        submit_result = submit_imagine(
            api_key=api_key,
            submit_url=submit_url,
            prompt=final_prompt,
            base64_array=base64_array,
            notify_hook=args.notify_hook.strip(),
            timeout=args.timeout,
        )
    except Exception as e:
        print(f"Submit failed: {e}", file=sys.stderr)
        return 1

    print(json.dumps(submit_result, ensure_ascii=False, indent=2))

    if args.no_poll:
        return 0

    task_id = str(submit_result.get("result") or "").strip()
    if not task_id:
        print("No task id found in submit result.result", file=sys.stderr)
        return 1

    fetch_url = build_fetch_url(base_url, route_prefix, task_id)
    print(f"Polling task: {task_id}", file=sys.stderr)
    try:
        task_result = poll_task(
            api_key=api_key,
            fetch_url=fetch_url,
            timeout=args.timeout,
            poll_interval=args.poll_interval,
            poll_timeout=args.poll_timeout,
        )
    except Exception as e:
        print(f"Polling failed: {e}", file=sys.stderr)
        return 1

    print(json.dumps(task_result, ensure_ascii=False, indent=2))

    if args.download:
        image_url = str(task_result.get("imageUrl") or "").strip()
        if not image_url:
            print("Task finished but imageUrl is empty", file=sys.stderr)
            return 1
        try:
            download_image(image_url, args.download, timeout=args.timeout)
            print(f"Downloaded image -> {args.download}", file=sys.stderr)
        except Exception as e:
            print(f"Download failed: {e}", file=sys.stderr)
            return 1

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/image-generation/scripts/mj_imagine_client.js
#!/usr/bin/env node
const path = require("node:path");
const { parseArgs } = require("node:util");
const {
  downloadFile,
  encodeImageToDataUri,
  isDataUri,
  isRemoteUrl,
  mergePromptWithPositionals,
  normalizeArgvForMultiValueOptions,
  parseRetryOptions,
  requestJson,
  resolveConfigPath,
  resolveProviderRuntime,
} = require("./lib/image_api_common");

const DEFAULT_BASE_URL = "https://api.bltcy.top";
const DEFAULT_MJ_VERSION = "7";

function resolveBaseUrl(values, config) {
  const raw = values["base-url"] || config.baseUrl || config.base_url || DEFAULT_BASE_URL;
  const finalValue = String(raw || "").trim().replace(/\/+$/, "");
  if (!finalValue) {
    throw new Error("Base URL is empty");
  }
  return finalValue;
}

function resolveRoutePrefix(values, config) {
  const raw = values["route-prefix"] !== undefined ? values["route-prefix"] : config.route_prefix || "fast";
  const text = String(raw || "").trim();
  return text ? text.replace(/^\/+|\/+$/g, "") : "";
}

function buildSubmitUrl(baseUrl, routePrefix) {
  return routePrefix ? `${baseUrl}/${routePrefix}/mj/submit/imagine` : `${baseUrl}/mj/submit/imagine`;
}

function buildFetchUrl(baseUrl, routePrefix, taskId) {
  return routePrefix ? `${baseUrl}/${routePrefix}/mj/task/${taskId}/fetch` : `${baseUrl}/mj/task/${taskId}/fetch`;
}

function normalizePrompt(prompt) {
  const text = String(prompt || "").trim();
  if (!text) {
    return text;
  }
  const lower = ` ${text.toLowerCase()} `;
  if (lower.includes(" --v ") || lower.includes(" --version ")) {
    return text;
  }
  return `${text} --v ${DEFAULT_MJ_VERSION}`;
}

function getImageInputs(values) {
  const raw = values["image-path"];
  if (!raw) {
    return [];
  }
  return Array.isArray(raw) ? raw : [raw];
}

function buildBase64Array(inputs) {
  const list = [];
  for (const item of inputs) {
    const text = String(item || "").trim();
    if (!text) {
      continue;
    }
    if (isRemoteUrl(text) || isDataUri(text)) {
      list.push(text);
    } else {
      list.push(encodeImageToDataUri(text));
    }
  }
  return list;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function buildIndexedPath(basePath, index, prefix = "") {
  const ext = path.extname(basePath);
  const stem = ext ? basePath.slice(0, -ext.length) : basePath;
  return `${stem}${prefix}${index}${ext}`;
}

function getSingleImageUrls(taskResult) {
  const raw = Array.isArray(taskResult.imageUrls) ? taskResult.imageUrls : [];
  const urls = [];
  for (const item of raw) {
    if (!item) {
      continue;
    }
    if (typeof item === "string") {
      const value = item.trim();
      if (value) {
        urls.push(value);
      }
      continue;
    }
    const value = String(item.url || "").trim();
    if (value) {
      urls.push(value);
    }
  }
  return urls;
}

async function downloadTaskImages(taskResult, downloadPath, downloadMode, timeoutSeconds, retry, retryDelay) {
  const mode = String(downloadMode || "grid").trim().toLowerCase();
  if (!["grid", "single", "both"].includes(mode)) {
    throw new Error("Invalid --download-mode. Use: grid, single, or both");
  }

  let downloadCount = 0;

  if (mode === "grid" || mode === "both") {
    const imageUrl = String(taskResult.imageUrl || "").trim();
    if (imageUrl) {
      await downloadFile(imageUrl, downloadPath, timeoutSeconds, retry, retryDelay);
      console.error(`Downloaded image (grid) -> ${downloadPath}`);
      downloadCount += 1;
    } else {
      console.error("Warning: imageUrl is empty, skip grid download");
    }
  }

  if (mode === "single" || mode === "both") {
    const urls = getSingleImageUrls(taskResult);
    if (urls.length === 0) {
      console.error("Warning: imageUrls is empty, skip single-image download");
    }
    for (let i = 0; i < urls.length; i += 1) {
      const savePath = urls.length === 1 ? downloadPath : buildIndexedPath(downloadPath, i + 1, "_single_");
      await downloadFile(urls[i], savePath, timeoutSeconds, retry, retryDelay);
      console.error(`Downloaded image (single) -> ${savePath}`);
      downloadCount += 1;
    }
  }

  if (downloadCount === 0) {
    throw new Error("Task finished but no downloadable image URL found");
  }
}

async function pollTask(apiKey, fetchUrl, timeoutSeconds, pollInterval, pollTimeout, retry, retryDelay) {
  const deadline = Date.now() + pollTimeout * 1000;
  let last = {};

  while (Date.now() < deadline) {
    last = await requestJson({
      method: "GET",
      url: fetchUrl,
      apiKey,
      body: undefined,
      timeoutSeconds,
      maxRetries: retry,
      retryDelayMs: retryDelay,
    });

    const status = String(last.status || "").toUpperCase();
    const progress = String(last.progress || "");
    if (["SUCCESS", "FAILURE", "CANCEL"].includes(status) || progress === "100%") {
      return last;
    }
    console.error(`Polling status: ${status || "PENDING"}${progress ? ` (${progress})` : ""}`);

    await sleep(Math.max(200, Math.floor(pollInterval * 1000)));
  }

  throw new Error(`Polling timed out after ${pollTimeout}s`);
}

async function main() {
  const scriptDir = __dirname;
  const configPath = resolveConfigPath(scriptDir);
  const normalizedArgv = normalizeArgvForMultiValueOptions(process.argv.slice(2), ["--image-path"]);

  let parsed;
  try {
    parsed = parseArgs({
      args: normalizedArgv,
      options: {
        prompt: { type: "string" },
        "image-path": { type: "string", multiple: true },
        "model-id": { type: "string", default: "" },
        "base-url": { type: "string", default: "" },
        "route-prefix": { type: "string" },
        "notify-hook": { type: "string", default: "" },
        timeout: { type: "string", default: "120" },
        retry: { type: "string", default: "2" },
        "retry-delay": { type: "string", default: "800" },
        "no-poll": { type: "boolean", default: false },
        "poll-interval": { type: "string", default: "3" },
        "poll-timeout": { type: "string", default: "600" },
        download: { type: "string", default: "" },
        "download-mode": { type: "string", default: "grid" },
        "dry-run": { type: "boolean", default: false },
      },
      strict: true,
      allowPositionals: true,
    });
  } catch (error) {
    console.error(`Argument parsing error: ${error.message}`);
    if (/unexpected argument/i.test(String(error.message || ""))) {
      console.error("PowerShell tip: use single quotes for prompts with spaces, e.g. --prompt 'A B C'.");
    }
    return 2;
  }

  const { values, positionals } = parsed;
  const promptResult = mergePromptWithPositionals(values.prompt, positionals);
  const prompt = promptResult.prompt;
  if (promptResult.merged) {
    console.error("Detected extra positional arguments. Auto-merged into --prompt.");
    console.error("PowerShell tip: use single quotes for prompts with spaces, e.g. --prompt 'A B C'.");
  }

  if (!prompt) {
    console.error("Missing required argument: --prompt");
    return 2;
  }

  const timeout = Number.parseInt(values.timeout, 10);
  const pollInterval = Number.parseFloat(values["poll-interval"]);
  const pollTimeout = Number.parseInt(values["poll-timeout"], 10);
  if (!Number.isFinite(timeout) || timeout <= 0) {
    console.error("Invalid --timeout value");
    return 2;
  }
  if (!Number.isFinite(pollInterval) || pollInterval <= 0) {
    console.error("Invalid --poll-interval value");
    return 2;
  }
  if (!Number.isFinite(pollTimeout) || pollTimeout <= 0) {
    console.error("Invalid --poll-timeout value");
    return 2;
  }
  let retryOptions;
  try {
    retryOptions = parseRetryOptions(values);
  } catch (error) {
    console.error(error.message);
    return 2;
  }
  const { retry, retryDelay } = retryOptions;
  if (!["grid", "single", "both"].includes(String(values["download-mode"] || "").trim().toLowerCase())) {
    console.error("Invalid --download-mode. Use: grid, single, or both");
    return 2;
  }

  let config;
  let apiKey;
  let baseUrl;
  let routePrefix;
  try {
    const runtime = resolveProviderRuntime(configPath, "mj", values["model-id"], DEFAULT_BASE_URL);
    config = runtime;
    apiKey = runtime.apiKey;
    baseUrl = resolveBaseUrl(values, runtime);
    routePrefix = resolveRoutePrefix(values, runtime);
  } catch (error) {
    console.error(`Config error: ${error.message}`);
    return 2;
  }

  let base64Array;
  try {
    base64Array = buildBase64Array(getImageInputs(values));
  } catch (error) {
    console.error(`Image error: ${error.message}`);
    return 2;
  }

  const submitUrl = buildSubmitUrl(baseUrl, routePrefix);
  const finalPrompt = normalizePrompt(prompt);

  console.error(`Submit URL: ${submitUrl}`);
  console.error(`Mode: ${base64Array.length > 0 ? "文图生图" : "文生图"}`);
  if (finalPrompt !== prompt) {
    console.error(`Auto append default MJ version: --v ${DEFAULT_MJ_VERSION}`);
  }

  const payload = {
    prompt: finalPrompt,
    base64Array,
  };
  if (String(values["notify-hook"] || "").trim()) {
    payload.notifyHook = String(values["notify-hook"]).trim();
  }

  if (values["dry-run"]) {
    console.log(JSON.stringify({
      ok: true,
      dry_run: true,
      submit_url: submitUrl,
      provider: "mj",
      model_id: config.modelId,
      route_prefix: routePrefix,
      payload,
      poll: !values["no-poll"],
    }, null, 2));
    return 0;
  }

  let submitResult;
  try {
    console.error("Submitting MJ task... please wait.");
    submitResult = await requestJson({
      method: "POST",
      url: submitUrl,
      apiKey,
      body: payload,
      timeoutSeconds: timeout,
      maxRetries: retry,
      retryDelayMs: retryDelay,
    });
  } catch (error) {
    console.error(`Submit failed: ${error.message}`);
    return 1;
  }

  console.log(JSON.stringify(submitResult, null, 2));

  if (values["no-poll"]) {
    return 0;
  }

  const taskId = String(submitResult.result || "").trim();
  if (!taskId) {
    console.error("No task id found in submit result.result");
    return 1;
  }

  const fetchUrl = buildFetchUrl(baseUrl, routePrefix, taskId);
  console.error(`Polling task: ${taskId}`);

  let taskResult;
  try {
    taskResult = await pollTask(apiKey, fetchUrl, timeout, pollInterval, pollTimeout, retry, retryDelay);
  } catch (error) {
    console.error(`Polling failed: ${error.message}`);
    return 1;
  }

  console.log(JSON.stringify(taskResult, null, 2));

  if (values.download) {
    try {
      await downloadTaskImages(
        taskResult,
        values.download,
        values["download-mode"],
        timeout,
        retry,
        retryDelay,
      );
    } catch (error) {
      console.error(`Download failed: ${error.message}`);
      return 1;
    }
  }

  return 0;
}

main()
  .then((code) => process.exit(code))
  .catch((error) => {
    console.error(error.message);
    process.exit(1);
  });
